commit cbcfeabace93906355a6cabf8f19dc944e8ad4d8
Author: Kevin Conner <kev.conner@getupcloud.com>
Date:   Thu Mar 28 08:54:52 2024 -0700

    upgrade base image packages

diff --git a/Dockerfile b/Dockerfile
index 9c7b2d83f..c63ca4e61 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,5 +1,6 @@
 FROM alpine:3.19.1
 RUN apk --no-cache add ca-certificates git
+RUN apk upgrade && rm /var/cache/apk/*
 COPY trivy /usr/local/bin/trivy
 COPY contrib/*.tpl contrib/
 ENTRYPOINT ["trivy"]
